benchmark/main.cpp

```diff
@@ -56,979 +50,1198 @@
+    mbedtls_aes_free(&aes);
+#endif /* MBEDTLS_AES_C && MBEDTLS_CIPHER_MODE_CBC */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_ctr() {
+    int ret = 0;
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CIPHER_MODE_CTR)
+    int keysize;
+    size_t nc_offset = 0;
+    unsigned char stream_block[16];
+    mbedtls_aes_context aes;
+
+    mbedtls_aes_init(&aes);
+
+    for ( keysize = 128; keysize <= 256; keysize += 64 ) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-CTR-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
         }
-        mbedtls_aes_free( &aes );
+
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        if ((ret = mbedtls_aes_setkey_enc(&aes, tmp, keysize)) != 0) {
+            mbedtls_printf("mbedtls_aes_setkey_enc() returned -0x%04X\n",
+                           -ret);
+            goto exit;
+        }
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_aes_crypt_ctr(&aes, BUFSIZE, &nc_offset,
+                                                  tmp, stream_block, buf,
+                                                  buf));
     }
-#endif
 
-#if defined(MBEDTLS_GCM_C)
-    if( todo->aes_gcm )
-    {
-        int keysize;
-        mbedtls_gcm_context gcm;
+exit:
+    mbedtls_aes_free(&aes);
+#endif /* MBEDTLS_AES_C && MBEDTLS_CIPHER_MODE_CTR */
+
+    return ret;
+}
+
+MBED_NOINLINE static int benchmark_aes_gcm() {
+    int ret = 0;
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_GCM_C)
+    int keysize;
+    mbedtls_gcm_context gcm;
 
-        mbedtls_gcm_init( &gcm );
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-GCM-%d", keysize );
+    mbedtls_gcm_init(&gcm);
 
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-            mbedtls_gcm_setkey( &gcm, MBEDTLS_CIPHER_ID_AES, tmp, keysize );
+    for (keysize = 128; keysize <= 256; keysize += 64) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-GCM-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
+        }
 
-            TIME_AND_TSC( title,
-                    mbedtls_gcm_crypt_and_tag( &gcm, MBEDTLS_GCM_ENCRYPT, BUFSIZE, tmp,
-                        12, NULL, 0, buf, buf, 16, tmp ) );
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
 
-            mbedtls_gcm_free( &gcm );
+        ret = mbedtls_gcm_setkey(&gcm, MBEDTLS_CIPHER_ID_AES, tmp, keysize);
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_gcm_setkey() returned -0x%04X\n", -ret);
```

Comment:
Please skip in case of feature not supported

Reply:
I think AES has the feature unavailable macro error code, but as far as I know GCM does not have it yet. I think the PR that adds it is: https://github.com/ARMmbed/mbedtls/pull/1716. Is this correct? If so, then I will not add the check because the symbol MBEDTLS_ERR_GCM_FEATURE_UNAVAILABLE does not exist yet. Or is it possible that this function returns MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE?

Reply:
Thanks for clarifying. I have added the checks for the AES feature unavailable only.